[JSC] Introduce toInt64 and toUInt64 with a double

hyjorc1 opened the pull request:
#### b22c0a8acec04b2e967aba07340fe8074982dbd2
<pre>
[JSC] Introduce toInt64 and toUInt64 with a double
<a href="https://bugs.webkit.org/show_bug.cgi?id=271015">https://bugs.webkit.org/show_bug.cgi?id=271015</a>
<a href="https://rdar.apple.com/124643241">rdar://124643241</a>

Reviewed by Yusuke Suzuki.

This patch extends the implementation of toInt32Internal for toInt64 and toUInt64.

* Source/JavaScriptCore/runtime/MathCommon.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/runtime/MathCommon.h:
(JSC::toIntImpl):
(JSC::toInt32):
(JSC::toInt64):
(JSC::toUInt64):
(JSC::toInt32Internal): Deleted.

Canonical link: <a href="https://commits.webkit.org/276173@main">https://commits.webkit.org/276173@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/41f813bf7bf4ea87aedb74dc959282e138bb8542

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43948 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23009 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46381 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46596 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40023 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26913 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20405 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44523 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20039 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37848 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17271 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/17547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38936 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2007 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/37324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40125 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39228 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48170 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/43541 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18954 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15526 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43099 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20342 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/41812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20547 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/50585 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6011 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19968 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10208 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/MathCommon.h:
```diff
@@ -80,51 +80,64 @@ inline bool isNegativeZero(double value)
+    constexpr unsigned intBits = intBytes * CHAR_BIT;
+    constexpr unsigned intBitsMinusOne = intBits - 1;
+    static_assert(intBitsMinusOne == 63 || intBitsMinusOne == 31);
+    using UInt = typename std::conditional<intBytes == 8, uint64_t, uint32_t>::type;
```
Use `using UInt = std::make_unsigned_t<Int>`

Constellation commented on Source/JavaScriptCore/runtime/MathCommon.h:
```diff
@@ -80,51 +80,64 @@ inline bool isNegativeZero(double value)
+    static constexpr bool value = std::is_same<Int, int32_t>::value || std::is_same<Int, int64_t>::value;
+};
+template<class Int, ToIntMode Mode = Generic>
+ALWAYS_INLINE constexpr typename std::enable_if<IsInt32OrInt64<Int>::value, Int>::type toIntImpl(double number)
```
We cannot use `constexpr` since `bitwise_cast` isn't `constexpr`.

Constellation commented on Source/JavaScriptCore/runtime/MathCommon.h:
```diff
@@ -80,51 +80,64 @@ inline bool isNegativeZero(double value)
+template<typename Int>
+struct IsInt32OrInt64 {
+    static constexpr bool value = std::is_same<Int, int32_t>::value || std::is_same<Int, int64_t>::value;
+};
```
Why not just check,

```
static_assert(std::is_same_v<Int, int32_t> || std::is_same_v<Int, int64_t>);
```

inside the function? It is simpler.

Constellation commented on Source/JavaScriptCore/runtime/MathCommon.h:
```diff
@@ -190,6 +203,17 @@ ALWAYS_INLINE constexpr UCPUStrictInt32 toUCPUStrictInt32(int32_t value)
+ALWAYS_INLINE int64_t toUInt64(double number)
+{
+    return static_cast<int64_t>(toInt64(number));
+}
```
It looks not correct, right? It returns `int64_t`, not `uint64_t`

Constellation reviewed (approved):
r=me

hyjorc1 committed:
[JSC] Introduce toInt64 and toUInt64 with a double
https://bugs.webkit.org/show_bug.cgi?id=271015
rdar://124643241

Reviewed by Yusuke Suzuki.

This patch extends the implementation of toInt32Internal for toInt64 and toUInt64.

* Source/JavaScriptCore/runtime/MathCommon.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/runtime/MathCommon.h:
(JSC::toIntImpl):
(JSC::toInt32):
(JSC::toInt64):
(JSC::toUInt64):
(JSC::toInt32Internal): Deleted.

Canonical link: https://commits.webkit.org/276173@main
